age/keysource.go

```diff
@@ -0,0 +1,196 @@
+package age
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"filippo.io/age"
+)
+
+// MasterKey is an age key used to encrypt and decrypt sops' data key.
+type MasterKey struct {
+	Identity     string // a Bech32-encoded private key
+	Recipient    string // a Bech32-encoded public key
+	EncryptedKey string // a sops data key encrypted with age
+
+	parsedIdentity  *age.X25519Identity  // a parsed age private key
+	parsedRecipient *age.X25519Recipient // a parsed age public key
+}
+
+// Encrypt takes a sops data key, encrypts it with age and stores the result in the EncryptedKey field.
+func (key *MasterKey) Encrypt(datakey []byte) error {
+	buffer := &bytes.Buffer{}
+
+	if key.parsedRecipient == nil {
+		parsedRecipient, err := age.ParseX25519Recipient(key.Recipient)
+
+		if err != nil {
+			return fmt.Errorf("failed to parse input as Bech32-encoded age public key: %v", err)
+		}
+
+		key.parsedRecipient = parsedRecipient
+	}
+
+	w, err := age.Encrypt(buffer, key.parsedRecipient)
+
+	if err != nil {
+		return fmt.Errorf("failed to open file for encrypting sops data key with age: %v", err)
+	}
+
+	if _, err := w.Write(datakey); err != nil {
+		return fmt.Errorf("failed to encrypt sops data key with age: %v", err)
+	}
+
+	if err := w.Close(); err != nil {
+		return fmt.Errorf("failed to close file for encrypting sops data key with age: %v", err)
+	}
+
+	key.EncryptedKey = buffer.String()
+
+	return nil
+}
+
+// EncryptIfNeeded encrypts the provided sops' data key and encrypts it if it hasn't been encrypted yet.
+func (key *MasterKey) EncryptIfNeeded(datakey []byte) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(datakey)
+	}
+
+	return nil
+}
+
+// EncryptedDataKey returns the encrypted data key this master key holds.
+func (key *MasterKey) EncryptedDataKey() []byte {
+	return []byte(key.EncryptedKey)
+}
+
+// SetEncryptedDataKey sets the encrypted data key for this master key.
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Decrypt decrypts the EncryptedKey field with the age identity and returns the result.
+func (key *MasterKey) Decrypt() ([]byte, error) {
+	ageKeyDir, ok := os.LookupEnv("SOPS_AGE_KEY_DIR")
+
+	if !ok {
+		userConfigDir, err := os.UserConfigDir()
+
+		if err != nil {
+			return nil, fmt.Errorf("user config directory could not be determined: %v", err)
+		}
+
+		ageKeyDir = filepath.Join(userConfigDir, ".sops", "age")
+	}
+
+	path := filepath.Join(ageKeyDir, fmt.Sprintf("%s.key", key.Recipient))
+
+	_, err := os.Stat(path)
+
+	if os.IsNotExist(err) {
+		return nil, fmt.Errorf("no private key found at %s", path)
+	}
+
+	file, err := os.Open(path)
+
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+
+	var privateKey string
+
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.HasPrefix(line, "AGE-SECRET-KEY") {
+			privateKey = line
+			break
+		}
+	}
```

Comment:
It would be nice to just copy this function: https://github.com/FiloSottile/age/blob/189041b668629795593766bcb8d3f70ee248b842/cmd/age/parse.go#L36

Then it would not just support the age format but also ssh keys like age does.
Same goes for encrypting.

Reply:
@FiloSottile How does making the linked code available via age's library sound? Ideally this logic wouldn't be duplicated in sops or it could get out of sync with what age does over time.

Reply:
That would be the ideal case.